### PR TITLE
Add feature flag `DisableDNSProviderManagement`

### PR DESCRIPTION
**Pull request by MartinWeindel**
**How to categorize this PR?**

/area control-plane
/kind task

**What this PR does / why we need it**:
The feature flag `DisableDNSProviderManagement` is introduced for the `gardenlet` to disable any handling of `DNSProvider`. It will only by effective if the feature flag `UseDNSRecords` is set to `true`. 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
Related to https://github.com/gardener/gardener/issues/5270

**Release note**:

```other operator
Add feature flag `DisableDNSProviderManagement`. This is part of [Move DNSProvider capabilities out of g/g #5270](https://github.com/gardener/gardener/issues/5270).
```

### CODE REVIEW

File: pkg/features/features.go [hunk 1]
```diff
@@ -129,4 +129,11 @@ const (
 	// owner: @stoyanr
 	// alpha: v1.39.0
 	ForceRestore featuregate.Feature = "ForceRestore"
+
+	// DisableDNSProviderMgmt disables management of DNSProviders for moving this responsibility to the shoot-dns-service.
+	// Only supported if feature `UseDNSRecords` is set to true.
+	// owner: @martinweindel @timuthy
+	// alpha: ?
+	// beta: ?
+	DisableDNSProviderMgmt featuregate.Feature = "DisableDNSProviderMgmt"
```
Comment (ialidzhikov) [inline pkg/features/features.go]: I would rather vote to have the full name such as `DisableDNSProviderManagement` - it is much more clear and we don't save much by using `Mgmt` over `Management`.

```suggestion
	DisableDNSProviderManagement featuregate.Feature = "DisableDNSProviderManagement"
```

File: pkg/features/features.go [hunk 2]
```diff
@@ -129,4 +129,11 @@ const (
 	// owner: @stoyanr
 	// alpha: v1.39.0
 	ForceRestore featuregate.Feature = "ForceRestore"
+
+	// DisableDNSProviderMgmt disables management of DNSProviders for moving this responsibility to the shoot-dns-service.
+	// Only supported if feature `UseDNSRecords` is set to true.
+	// owner: @martinweindel @timuthy
+	// alpha: ?
+	// beta: ?
```
Comment (ialidzhikov) [inline pkg/features/features.go]: This field is only populated when the feature gate is promoted to beta.
```suggestion
```

File: docs/deployment/feature_gates.md [hunk 3]
```diff
@@ -42,6 +42,7 @@ The following tables are a summary of the feature gates that you can set on diff
 | CopyEtcdBackupsDuringControlPlaneMigration   | `false` | `Alpha` | `1.37` |        |
 | SecretBindingProviderValidation              | `false` | `Alpha` | `1.38` |        |
 | ForceRestore                                 | `false` | `Alpha` | `1.39` |        |
+| DisableDNSProviderMgmt                       | `false` | `Alpha` |        |        |
```
Comment (ialidzhikov) [inline docs/deployment/feature_gates.md]: In the `List of feature gates` section, each feature gate is described with few sentences. You could describe the new feature gate in this section as well.

File: pkg/features/features.go [hunk 4]
```diff
@@ -129,4 +129,11 @@ const (
 	// owner: @stoyanr
 	// alpha: v1.39.0
 	ForceRestore featuregate.Feature = "ForceRestore"
+
+	// DisableDNSProviderMgmt disables management of DNSProviders for moving this responsibility to the shoot-dns-service.
+	// Only supported if feature `UseDNSRecords` is set to true.
+	// owner: @martinweindel @timuthy
```
Comment (ialidzhikov) [inline pkg/features/features.go]:
```suggestion
	// owner: @MartinWeindel @timuthy
```

File: pkg/operation/botanist/dns.go [hunk 5]
```diff
@@ -570,3 +578,7 @@ func (b *Botanist) newDNSComponentsTargetingAPIServerAddress() {
 		b.Shoot.Components.Extensions.ExternalDNSRecord.SetValues([]string{b.APIServerAddress})
 	}
 }
+
+func disabledDNSProviderMgmt() bool {
+	return gardenletfeatures.FeatureGate.Enabled(features.UseDNSRecords) && gardenletfeatures.FeatureGate.Enabled(features.DisableDNSProviderMgmt)
+}
```
Comment (ialidzhikov) [inline pkg/operation/botanist/dns.go]: This function is duplicated 2 times - it is probably a sign that it should be placed in a common place and used by callers from there.

File: pkg/operation/botanist/dns.go [hunk 6]
```diff
@@ -423,6 +425,9 @@ func (b *Botanist) DestroyInternalDNS(ctx context.Context) error {
 
 // DestroyExternalDNS destroys the external DNSEntry, DNSOwner, and DNSProvider resources.
 func (b *Botanist) DestroyExternalDNS(ctx context.Context) error {
+	if disabledDNSProviderMgmt() {
+		return nil
+	}
```
Comment (ialidzhikov) [inline pkg/operation/botanist/dns.go]: Unit tests can be added that verify that this func won't actually call Destroy for the resources below when the `DisableDNSProviderManagement` feature gate is enabled.

File: docs/deployment/feature_gates.md (hunk 3 above)
Comment (ialidzhikov) [inline docs/deployment/feature_gates.md]: You could also add the feature gate to the example component config (`example/20-componentconfig-gardenlet.yaml`).

File: pkg/features/features.go [hunk 7]
```diff
@@ -129,4 +129,11 @@ const (
 	// owner: @stoyanr
 	// alpha: v1.39.0
 	ForceRestore featuregate.Feature = "ForceRestore"
+
+	// DisableDNSProviderMgmt disables management of DNSProviders for moving this responsibility to the shoot-dns-service.
```
Comment (rfranzke) [inline pkg/features/features.go]:
```suggestion
	// DisableDNSProviderMgmt disables management of `dns.gardener.cloud/v1alpha1.DNSProvider` resources. In this case, the `shoot-dns-service` extension will take this over if it is installed.
```

File: pkg/features/features.go [hunk 8]
```diff
@@ -129,4 +129,11 @@ const (
...
+	// Only supported if feature `UseDNSRecords` is set to true.
+	// owner: @martinweindel @timuthy
+	// alpha: ?
```
Comment (rfranzke) [inline pkg/features/features.go]:
```suggestion
	// alpha: v1.41
```

File: docs/deployment/feature_gates.md (hunk 3 above)
Comment (MartinWeindel) [reply docs/deployment/feature_gates.md]: done

File: pkg/operation/botanist/dns.go (hunk 5 above)
Comment (MartinWeindel) [reply pkg/operation/botanist/dns.go]: moved to pkg/gardenlet/features/features.go

**Review by rfranzke — DISMISSED** [review]
/lgtm

File: pkg/operation/botanist/dns.go (hunk 6 above)
Comment (MartinWeindel) [reply pkg/operation/botanist/dns.go]: added unit test in [pkg/operation/botanist/dnsresources_test.go](https://github.com/gardener/gardener/pull/5349/commits/d58887e986164de2dfeba1384863cae83d7dbdc9#diff-78073b99419520088981083af7f99d96023ef92a2fd28182f60e5f72811c5740)

File: docs/deployment/feature_gates.md (hunk 3 above)
Comment (MartinWeindel) [reply docs/deployment/feature_gates.md]: done

File: pkg/features/features.go (hunk 7 above)
Comment (MartinWeindel) [reply pkg/features/features.go]: done

**Review by rfranzke — APPROVED** [review]

**Comment by timebertt** [comment]
@MartinWeindel as there is no documentation included in this PR, maybe it would be good to reference https://github.com/gardener/gardener/issues/5270 in the release note at least.

**Comment by MartinWeindel** [comment]
> @MartinWeindel as there is no documentation included in this PR, maybe it would be good to reference #5270 in the release note at least.

@timebertt  Done. Thanks for the hint.

**Pull request merged**